### PR TITLE
[ACCOUNT-2499] feat: remove config if not client id or secret

### PR DESCRIPTION
**Pull request by atourneriePresta**

### CODE REVIEW

File: tests/Unit/Account/Session/Oauth2/ShopSession/RefreshTokenTest.php [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+class RefreshTokenTest extends TestCase
+{
+    use \PrestaShop\Module\PsAccounts\Tests\Unit\Account\Session\SessionHelpers;
+
+    /**
+     * @inject
+     *
+     * @var ShopTokenRepository
+     */
+    protected $repository;
```
Comment (hschoenenberger) [inline tests/Unit/Account/Session/Oauth2/ShopSession/RefreshTokenTest.php:21]: the test should not rely on deprecated services

**Review by hschoenenberger — CHANGES_REQUESTED** [review]

File: src/Account/Session/ShopSession.php [hunk 2]
```diff
@@ -78,6 +85,11 @@ public function getOrRefreshToken($forceRefresh = false)
     public function refreshToken($refreshToken = null)
     {
         try {
+            if (!$this->oauth2ClientProvider->getOauth2Client()->exists()) {
+                $this->linkShop->delete();
+                $this->configurationRepository->clearLinkContext();
+                throw new RefreshTokenException('Oauth2 client id or Oauth secret is empty');
```
Comment (hschoenenberger) [inline src/Account/Session/ShopSession.php:91]:
```suggestion
                throw new RefreshTokenException('Invalid OAuth2 client');
```

File: src/Account/Session/ShopSession.php [hunk 3]
```diff
@@ -78,6 +85,11 @@ public function getOrRefreshToken($forceRefresh = false)
     public function refreshToken($refreshToken = null)
     {
         try {
+            if (!$this->oauth2ClientProvider->getOauth2Client()->exists()) {
+                $this->linkShop->delete();
+                $this->configurationRepository->clearLinkContext();
```
Comment (hschoenenberger) [inline src/Account/Session/ShopSession.php:90]: maybe we should trigger an UnlinkShopCommand here which already do the local cleanup the right way.

File: tests/Unit/Account/Session/Oauth2/ShopSession/RefreshTokenTest.php [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+        $this->createMockedSession(true, null);
+
+        $idToken = $this->makeJwtToken(new \DateTimeImmutable('yesterday'), [
+            'sub' => $this->faker->uuid,
+        ]);
+        $refreshToken = $this->makeJwtToken(new \DateTimeImmutable('+1 year'));
+        $this->repository->updateCredentials((string) $idToken, (string) $refreshToken);
+        try {
+            $this->shopSession->refreshToken();
+        } catch (\Exception $e) {
```
[the remark this answers is not in the document]
Comment (atourneriePresta) [reply tests/Unit/Account/Session/Oauth2/ShopSession/RefreshTokenTest.php:46]: Yes i could, but i found it more readable this way + i can add a fail at the end of the test if it doesn't throw

File: tests/Unit/Account/Session/Oauth2/ShopSession/RefreshTokenTest.php (hunk 1 above)
Comment (atourneriePresta) [reply tests/Unit/Account/Session/Oauth2/ShopSession/RefreshTokenTest.php:21]: Actually, i have to check if the token is still present in the db after the test is done.
What's your idea on how to check without the repository?

File: src/Account/Session/ShopSession.php (hunk 3 above)
Comment (atourneriePresta) [reply src/Account/Session/ShopSession.php:90]: Yes actually we though about it during the pair coding but you said there wasn't a clean of everything like in the description of the issue i think :thinking:

File: Makefile [hunk 5]
```diff
@@ -99,7 +99,7 @@ phpunit-permissions:
 	@docker exec phpunit sh -c "if [ -d ./log ]; then chown -R www-data:www-data ./log; fi" # PS1.6
 
 phpunit-run-unit: phpunit-permissions vendor-dev
-	@docker exec -w ${CONTAINER_INSTALL_DIR} phpunit ./vendor/bin/phpunit --testsuite unit
+	@docker exec -w ${CONTAINER_INSTALL_DIR} phpunit ./vendor/bin/phpunit --testsuite unit --process-isolation
```
[the remark this answers is not in the document]
Comment (atourneriePresta) [reply Makefile:102]: For the test to pass it is + we won't have any more problems for futures tests.
It doesn't cost anything and solve the memory problem we could add

[the remark this answers is not in the document]
Comment (hschoenenberger) [reply Makefile:102]: worth looking further to investigate before applying

File: tests/Unit/Account/Session/Oauth2/ShopSession/RefreshTokenTest.php (hunk 1 above)
Comment (hschoenenberger) [reply tests/Unit/Account/Session/Oauth2/ShopSession/RefreshTokenTest.php:21]: `Firebase\ShopSession` is the replacement

File: src/Account/Session/ShopSession.php (hunk 3 above)
Comment (hschoenenberger) [reply src/Account/Session/ShopSession.php:90]: after looking further that's not the case, the command cleanup link context triggering a dedicated hook

Comment (atourneriePresta) [reply src/Account/Session/ShopSession.php:90]: Okay, i should so change all the code written and the test so?

Comment (hschoenenberger) [reply src/Account/Session/ShopSession.php:90]: yes, something like that should do the job :
`$this->commandBus->handle(new UnlinkShopCommand($shop->id));`

File: src/Repository/ConfigurationRepository.php [hunk 6]
```diff
@@ -373,6 +373,19 @@ public function isMultishopActive()
...
+     * @return void
+     */
+    public function clearLinkContext()
+    {
+        $this->updateFirebaseIdAndRefreshTokens('', '');
+        $this->updateUserFirebaseIdAndRefreshToken('', '');
+        $this->updateOauth2ClientId('');
+        $this->updateOauth2ClientSecret('');
+    }
+
```
Comment (hschoenenberger) [inline src/Repository/ConfigurationRepository.php:388]: this method can be removed

File: src/Account/Session/ShopSession.php [hunk 7]
```diff
@@ -35,25 +37,41 @@
...
 
+    /**
+     * @var ShopProvider
+     */
+    protected $oauth2ClientProvider;
+
+    /**
+     * @var LinkShop
+     */
+    protected $linkShop;
```
Comment (hschoenenberger) [inline src/Account/Session/ShopSession.php:57]: apparently not used anymore, but the `getShopUuid` method is probably useless if we use this member instead.

File: tests/Feature/Api/v1/ShopLinkAccount/StoreTest.php [hunk 8]
```diff
@@ -15,7 +15,7 @@ class StoreTest extends FeatureTestCase
      *
      * @var LinkShop
      */
-    protected $linkShop;
+    public $linkShop;
```
Comment (hschoenenberger) [inline tests/Feature/Api/v1/ShopLinkAccount/StoreTest.php:18]: same as above

File: tests/Feature/Api/v1/ShopLinkAccount/DeleteTest.php [hunk 9]
```diff
@@ -17,7 +17,7 @@ class DeleteTest extends FeatureTestCase
      *
      * @var LinkShop
      */
-    protected $linkShop;
+    public $linkShop;
```
Comment (hschoenenberger) [inline tests/Feature/Api/v1/ShopLinkAccount/DeleteTest.php:20]: what's the point injecting it again ?

File: tests/TestCase.php [hunk 10]
```diff
@@ -192,4 +204,19 @@ protected function createMockWithMethods($class, $methods = [])
...
     }
+
+    /**
+     * @param $classInstance
+     * @param $dependencyName
+     * @param $newDependency
+     * @return void
+     * @throws \ReflectionException
+     */
+    protected function replaceDependency($classInstance, $dependencyName, $newDependency)
```
Comment (hschoenenberger) [inline tests/TestCase.php:215]: Very usefull I think, let's see if we can refactor our tests with it in the future :+1:

**Review by hschoenenberger — CHANGES_REQUESTED** [review]

**Review by hschoenenberger — APPROVED** [review]

**Pull request merged**